File: 0907-sum-of-subarray-minimums/0907-sum-of-subarray-minimums.cpp
class Solution {
public:
    vector<int> findNSE(vector<int>& arr) {
        stack<int> st;
        vector<int> ans(arr.size());
        for(int i=arr.size() - 1; i>=0; i--) {
            while(!st.empty() && arr[st.top()] >= arr[i]) {
                st.pop();
            }
            ans[i] = st.empty() ? arr.size() : st.top();
            st.push(i);
        }
        return ans;
    }

    vector<int> findPSEE(vector<int>& arr) {
        stack<int> st;
        vector<int> ans(arr.size());
        for(int i=0; i<arr.size(); i++) {
            while(!st.empty() && arr[st.top()] > arr[i]) {
                st.pop();
            }
            ans[i] = st.empty() ? -1 : st.top();
            st.push(i);
        }
        return ans;
    }
    int sumSubarrayMins(vector<int>& arr) {
        vector<int> nse = findNSE(arr);
        vector<int> psee = findPSEE(arr);
        int mod = 1e9 + 7;
        int total = 0;
        for(int i=0; i<arr.size(); i++) {
            int left = i - psee[i];
            int right = nse[i] - i;

            total = ((total + (left*right) * 1LL * arr[i]) % mod) % mod;

        }
        return total;
    }
};

File: 0907-sum-of-subarray-minimums/0907-sum-of-subarray-minimums.java
class Solution {
    static int[] findNSE(int[] arr) {
        int n = arr.length;

        int[] res = new int[n];

        Stack<Integer> st = new Stack<>();
        for(int i=n-1; i>=0; i--) {
            int curr = arr[i];

            while(!st.isEmpty() && arr[st.peek()] >= curr) {
                st.pop();
            }
            res[i] = st.isEmpty() ? n : st.peek();
            st.push(i);
        }
        return res;
    }
    static int[] findPSE(int[] arr) {
        int n = arr.length;

        Stack<Integer> st = new Stack<>();
        int[] res = new int[n];

        for(int i=0; i<n; i++) {
            int curr = arr[i];

            while(!st.isEmpty() && arr[st.peek()] > curr) {
                st.pop();
            }

            res[i] = st.isEmpty() ? -1 : st.peek();
            st.push(i);
        }
        return res;
    }
    public int sumSubarrayMins(int[] arr) {
        long sum = 0;
        int mod = (int)(1e9 + 7);

        int n = arr.length;

        int[] pse = findPSE(arr);
        int[] nse = findNSE(arr);

        for(int i=0; i<n; i++) {
            long left = i - pse[i];
            long right = nse[i] - i;

            long totalSubarray = (left * right) % mod;
            sum = (sum + (totalSubarray * arr[i]) % mod) % mod;
        }
        return (int)sum;
    }
}